pkg/controller/kogitoapp/kogitoapp_controller.go

```diff
@@ -220,3 +230,91 @@ func (r *ReconcileKogitoApp) setDefaultBuildLimits(instance *v1alpha1.KogitoApp)
+
+// Ensure that all Kogito images are available before the build.
+func (r *ReconcileKogitoApp) ensureKogitoImageStream(instance *v1alpha1.KogitoApp) (reconcile.Result, error) {
+	need2createIS := false
+
+	imageStreamTag := kogitores.ImageStreamTag
+	if instance.Spec.Build.ImageS2I.ImageStreamTag != "" {
+		imageStreamTag = instance.Spec.Build.ImageS2I.ImageStreamTag
+		log.Infof("Using ImageS2I Tag from CR: %s", imageStreamTag)
+	}
+
+	// look for the Kogito image stream on the default namespace (openshift)
+	// All images must be present, if not create it on current namespace
+	for _, key := range kogitores.ImageStreamNameList {
+		isOnOCPNamespace, _ := openshift.ImageStreamC(r.client).FetchTag(types.NamespacedName{Name: key, Namespace: kogitores.ImageStreamNamespace}, imageStreamTag)
+		if isOnOCPNamespace == nil {
+			need2createIS = true
+		}
+	}
+
+	if need2createIS {
+		imageStreamLocation := fmt.Sprintf(kogitores.ImageStreamLocation, imageStreamTag)
+		log.Infof("Using imagestream %s", imageStreamLocation)
+		resp, err := http.Get(imageStreamLocation)
+		if err != nil {
+			log.Errorf("Failed to fetch imagestream %s", err.Error())
+			return reconcile.Result{}, err
+		}
+		defer resp.Body.Close()
+
+		if resp.StatusCode == http.StatusOK {
+
+			var imagestreamlist oimagev1.ImageStreamList
+			bodyBytes, err := ioutil.ReadAll(resp.Body)
+			if err != nil {
+				log.Warnf("Failed to read response: %s", err.Error())
+			}
+
+			if err := yaml.NewYAMLOrJSONDecoder(bytes.NewReader(bodyBytes), len([]byte(bodyBytes))).Decode(&imagestreamlist); err != nil {
+				return reconcile.Result{}, fmt.Errorf("Error while unmarshal file '%s': %s ", resp.Body, err)
+			}
+
+			for _, is := range imagestreamlist.Items {
+				is.Namespace = instance.Namespace
+				hasIs, err := openshift.ImageStreamC(r.client).FetchTag(types.NamespacedName{Name: is.Name, Namespace: instance.Namespace}, imageStreamTag)
+				if err != nil {
+					log.Error(err.Error())
+				}
+
+				if hasIs == nil {
+					_, err := openshift.ImageStreamC(r.client).CreateImageStream(&is)
+					if err != nil {
+						return reconcile.Result{}, err
+					}
+
+					// wait the image get ready, otherwise no image will be found when the build start
+					tagRef := fmt.Sprintf("%s:%s", is.Name, imageStreamTag)
+					counter := 0
+					for {
+						//imageTag, _ := r.client.ImageCli.ImageStreamTags(is.Namespace).Get(tagRef, metav1.GetOptions{})
+						imageTag, err1 := openshift.ImageStreamC(r.client).FetchTag(types.NamespacedName{Name: is.Name, Namespace: is.Namespace}, imageStreamTag)
+						if err1 != nil {
+							log.Error(err1)
+						}
+						if imageTag != nil && imageTag.Image.Name != "" {
+							log.Infof("Image %s ready.", imageTag.Name)
+							break
+						}
+						if counter >= 10 {
+							return reconcile.Result{}, fmt.Errorf("Failed to get information about image tag ref %s on namespace %s", tagRef, is.Namespace)
+						}
+						counter++
+						time.Sleep(500 * time.Microsecond)
```

Comment:
Isn't 500 microseconds too small step? If I count correctly the loop will wait for 5 milliseconds to create tag, then it returns error.

Reply:
not really, it wait for 5 seconds in total.
note that if the counter is greater or equal to 10 if returns a error.